app/mailers/notification_mailer.rb

```diff
@@ -0,0 +1,39 @@
+class NotificationMailer < ActionMailer::Base
+  INSTANT_DIGEST_INTERVAL = 10.minutes
+
+  default from: 'email@securityroots.com'
+  default to: 'admin@securityroots.com'
+
+  def digest(user:)
+    @user = user
+    @notifications = digest_notifications
+
+    @user.last_digest_email = Time.now
+    @user.save
+
+    mail to: user.email, subject: 'Digest Email'
+  end
+
+  def instant(user:)
+    @user = user
+    @notifications = instant_notifications
+
+    mail to: user.email, subject: 'Notifications in the last 10 minutes.'
+  end
+
+  private
+
+  def digest_notifications
```

Comment:
@aapomm I've got a proposal.

We could define a class `Digest` that the thor tasks can call `Digest.send` or `Digest.send_instant`.

This class can manage finding, creating notification groups, and calling the mailer with the collection.

This can make the mailer "dumb" about the collection is receives. And moves the collection generation to an easier to test class, that is more accessible to our thor tasks.

Reply:
Here's an example of what I mean (not final), and can likely my optimized/smaller, opinions welcome.

```ruby
class Digest
  INSTANT_DIGEST_INTERVAL = 10.minutes

  def self.send
    User.all.each do |user|
      Digest.new(user).send
    end
  end

  def self.send_instant
    User.all.each do |user|
      Digest.new(user).send_instant
    end
  end

  def initialize(user)
    @user = user
  end

  attr_accessor :user

  def send
    notifications = user.notifications.
      where('created_at >= ?', user.last_digest_email).
      unread.
      newest

    NotificationMailer.with(user: user, notifications: notifications).digest
  end

  def send_isntant
    notifications = user.notifications.
      where('created_at >= ?', Time.now - INSTANT_DIGEST_INTERVAL).
      unread.
      newest

    NotificationMailer.with(user: user, notifications: notifications).instant
  end
end
```

Reply:
Implemented this wrapper and went with `DigestMailer` since `Digest` is already being used. Also, thor will call the class methods `DigestMailer.send_digests` or `DigestMailer.send_instants`. I don't think thor should concern itself with the logic of which users to send it to.